crates/cranelift/src/gc/enabled.rs

```diff
@@ -826,6 +826,237 @@ pub fn translate_array_set(
     write_field_at_addr(func_env, builder, elem_ty, elem_addr, value)
 }
 
+pub fn translate_ref_test(
+    func_env: &mut FuncEnvironment<'_>,
+    builder: &mut FunctionBuilder<'_>,
+    ref_ty: WasmRefType,
+    val: ir::Value,
+) -> WasmResult<ir::Value> {
+    use crate::translate::FuncEnvironment as _;
+
+    // First special case: testing for references to bottom types.
+    if ref_ty.heap_type.is_bottom() {
+        let result = if ref_ty.nullable {
+            // All null references (within the same type hierarchy) match null
+            // references to the bottom type.
+            func_env.translate_ref_is_null(builder.cursor(), val)?
+        } else {
+            // `ref.test` is always false for non-nullable bottom types, as the
+            // bottom types are uninhabited.
+            builder.ins().iconst(ir::types::I32, 0)
+        };
+        return Ok(result);
+    }
+
+    // And because `ref.test heap_ty` is only valid on operands whose type is in
+    // the same type hierarchy as `heap_ty`, if `heap_ty` is its hierarchy's top
+    // type, we only need to worry about whether we are testing for nullability
+    // or not.
+    if ref_ty.heap_type.is_top() {
+        let result = if ref_ty.nullable {
+            builder.ins().iconst(ir::types::I32, 1)
+        } else {
+            let is_null = func_env.translate_ref_is_null(builder.cursor(), val)?;
+            let zero = builder.ins().iconst(ir::types::I32, 0);
+            let one = builder.ins().iconst(ir::types::I32, 1);
+            builder.ins().select(is_null, zero, one)
+        };
+        return Ok(result);
+    }
+
+    // `i31ref`s are a little interesting because they don't point to GC
+    // objects; we test the bit pattern of the reference itself.
+    if ref_ty.heap_type == WasmHeapType::I31 {
+        let i31_mask = builder.ins().iconst(
+            ir::types::I32,
+            i64::try_from(wasmtime_environ::I31_DISCRIMINANT).unwrap(),
+        );
+        let is_i31 = builder.ins().band(val, i31_mask);
+        let result = if ref_ty.nullable {
+            let is_null = func_env.translate_ref_is_null(builder.cursor(), val)?;
+            builder.ins().bor(is_null, is_i31)
+        } else {
+            is_i31
+        };
+        return Ok(result);
+    }
+
+    // Otherwise, in the general case, we need to inspect our given object's
+    // actual type, which also requires null-checking and i31-checking it.
+
+    let is_any_hierarchy = matches!(
```

Comment:
Could this be replaced with `.top_type() == Something`?

Reply:
ah yes, good call